// @sap/hdi-deploy/lib/make.js
'use strict';
const hana_helper = require('./hana-helper.js');
const logger = require('./logger');
const connections = require('./connections');
const utils = require('./utils.js');
const { identifier } = utils;

const printMessages = hana_helper().printMessages;
/**
 * Log the messages.
 *
 * @param {any} messages The messages to log.
 * @param {any} logger Logging utility.
 * @returns {undefined} No return value.
 */
function logfn (messages) {
  printMessages(logger, messages);
}
/**
 * Class that handles the polling for make messages.
 *
 * @class Poller
 */
class Poller {
  /**
   * Creates an instance of Poller.
   *
   * @memberOf Poller
   */
  constructor (connection_id, schema, db_credentials) {
    this.last_row_id = 0;
    this.stop_flag = false;

    /**
     * Get the request_id by using the connection_id in M_JOBS. Need to poll, since we don't know, when the make has "hit the db".
     *
     * @param {any} connection_id
     * @param {any} schema
     * @param {any} db
     * @param {any} cb
     * @returns
     */
    const poll_request_id = (connection_id, schema, db, cb) => {
      if (this.stop_flag) {
        return cb();
      }
      const diSchema = identifier(`${schema}#DI`);
      const sql = `SELECT REQUEST_ID FROM ${diSchema}.M_JOBS WHERE CONNECTION_ID=${connection_id};`;
      /**
       * Callback for tail recursion
       *
       * @param {any} err
       * @param {any} results
       * @returns
       */
      const poll_callback = (err, results) => {
        if (this.stop_flag) {
          return cb();
        }
        if (err) {
          this.stop_flag = true;
          logger.warn('Encountered an error during polling for messages. Polling will be stopped.');
          return cb();
        }

        if (results && results.length === 1) {
          return cb(results[0].REQUEST_ID);
        }

        return setTimeout(() => db.exec(sql, poll_callback), 500);
      };

      db.exec(sql, poll_callback);
    };

    /**
     * Poll the messages for the currently running make
     *
     * @param {any} request_id
     * @param {any} schema
     * @param {any} db
     */
    const poll_messages = (request_id, schema, db) => {
      logger.log(`Polling messages for request id: ${request_id}`);
      if (this.stop_flag) {
        return;
      }
      request_id = parseInt(request_id);
      let row_id = -1;
      let statement;

      /**
       * Poll the messages and print them - Calls itself recursively
       */
      const poll_callback = (err, result_messages) => {
        if (this.stop_flag) {
          return;
        }
        if (err) {
          this.stop_flag = true;
          logger.warn('Encountered an error during polling for messages. Polling will be stopped.');
          return;
        }

        if (result_messages && result_messages.length) {
          logfn(result_messages);
          row_id = result_messages[result_messages.length - 1].ROW_ID;
          this.last_row_id = row_id;
        }
        if (logger.getClientValue() === logger.hdb) {
          return setTimeout(() => statement.exec([request_id, row_id], {}, poll_callback), 1000);
        } else {
          return setTimeout(() => statement.exec([request_id, row_id], poll_callback), 1000);
        }
      };

      const diSchema = identifier(`${schema}#DI`);
      const sql = `SELECT * FROM ${diSchema}.M_MESSAGES WHERE REQUEST_ID = ? AND ROW_ID > ? ORDER BY ROW_ID;`;
      db.prepare(sql, (err, prep_statement) => {
        if (err) {
          return poll_callback(err);
        }
        statement = prep_statement;
        statement.exec([request_id, row_id], poll_callback);
      });
    };

    this.start_polling = () => {
      if (!connection_id) {
        // Return, since without connection_id, no polling is possible.
        return;
      }
      try {
        let db;
        if (logger.getClientValue() === logger.hdb) {
          db = utils.getClient().createClient(db_credentials);
          connections.push({client:db, file: __filename});
          db.connect((e) => {
            if (e) {
              this.stop_flag = true;
              logger.warn('Encountered an error during polling for messages. Polling will be stopped.');
            } else {
              this.db = db;
              poll_request_id(connection_id, schema, db, (request_id) => {
                if (request_id) {
                  poll_messages(request_id, schema, db);
                }
              });
            }
          });
        } else {
          db = utils.getClient().createConnection();
          connections.push({client:db, file: __filename});
          db.connect(db_credentials, (e) => {
            if (e) {
              this.stop_flag = true;
              logger.warn('Encountered an error during polling for messages. Polling will be stopped.');
            } else {
              this.db = db;
              poll_request_id(connection_id, schema, db, (request_id) => {
                if (request_id) {
                  poll_messages(request_id, schema, db);
                }
              });
            }
          });
        }
      } catch (e) {
        this.stop_flag = true;
        logger.warn('Encountered an error during polling for messages. Polling will be stopped.');
      }
    };
  }

  /**
   * Stop the polling and return the last row id.
   *
   * @returns {String} Last row id
   *
   * @memberOf Poller
   */
  stop_polling () {
    this.stop_flag = true;
    if (this.db) {
      this.db.disconnect();
    }
    return this.last_row_id;
  }
}

const TERMINAL_SUCCESS = new Set(['FINISHED', 'COMMITTED', 'SIMULATED']);
const TERMINAL_ROLLBACK = new Set(['ROLLED BACK']);

/**
 * Polls M_MESSAGES and M_JOBS on a secondary connection until the async make job completes.
 * Used when --async-make is enabled: the MAKE_ASYNC call returns quickly with a requestId,
 * and this poller waits for the job to finish while streaming messages.
 *
 * @class AsyncPoller
 */
class AsyncPoller {
  constructor (schema, db_credentials, timeout) {
    this.schema = schema;
    this.db_credentials = db_credentials;
    this.timeout = timeout;
    this.last_row_id = 0;
    this.all_messages = [];
    this.db = null;
    this.connEntry = null;
  }

  /**
   * Start polling for the given requestId. Calls onDone when the job finishes.
   *
   * @param {number} requestId
   * @param {Function} onDone called with (err, {rc, messages, last_row_id})
   */
  poll (requestId, onDone) {
    requestId = parseInt(requestId, 10);
    if (isNaN(requestId)) {
      return onDone(new Error('MAKE_ASYNC did not return a valid requestId'));
    }

    const diSchema = identifier(`${this.schema}#DI`);
    const isHdb = logger.getClientValue() === logger.hdb;

    // Wrap onDone so it fires at most once and clears the timeout guard
    let settled = false;
    // eslint-disable-next-line prefer-const
    let timeoutId;
    const finish = (err, result) => {
      if (settled) return;
      settled = true;
      clearTimeout(timeoutId);
      onDone(err, result);
    };

    timeoutId = setTimeout(() => {
      if (this.db) this.db.disconnect();
      const idx = connections.indexOf(this.connEntry);
      if (idx !== -1) connections.splice(idx, 1);
      finish(new Error(`Async make polling timed out after ${this.timeout}ms`));
    }, this.timeout);

    try {
      let db;
      if (isHdb) {
        db = utils.getClient().createClient(this.db_credentials);
        this.connEntry = {client: db, file: __filename};
        connections.push(this.connEntry);
        db.connect((e) => {
          if (e) {
            const idx = connections.indexOf(this.connEntry);
            if (idx !== -1) connections.splice(idx, 1);
            return finish(e);
          }
          this.db = db;
          this._startPolling(requestId, diSchema, isHdb, finish);
        });
      } else {
        db = utils.getClient().createConnection();
        this.connEntry = {client: db, file: __filename};
        connections.push(this.connEntry);
        db.connect(this.db_credentials, (e) => {
          if (e) {
            const idx = connections.indexOf(this.connEntry);
            if (idx !== -1) connections.splice(idx, 1);
            return finish(e);
          }
          this.db = db;
          this._startPolling(requestId, diSchema, isHdb, finish);
        });
      }
    } catch (e) {
      if (this.connEntry) {
        const idx = connections.indexOf(this.connEntry);
        if (idx !== -1) connections.splice(idx, 1);
      }
      finish(e);
    }
  }

  _startPolling (requestId, diSchema, isHdb, onDone) {
    const db = this.db;
    const removeConn = () => {
      const idx = connections.indexOf(this.connEntry);
      if (idx !== -1) connections.splice(idx, 1);
    };
    const messageSql = `SELECT * FROM ${diSchema}.M_MESSAGES WHERE REQUEST_ID = ? AND ROW_ID > ? ORDER BY ROW_ID;`;

    db.prepare(messageSql, (err, msgStatement) => {
      if (err) {
        db.disconnect();
        removeConn();
        return onDone(err);
      }

      const jobSql = `SELECT STATUS FROM ${diSchema}.M_JOBS WHERE REQUEST_ID = ?;`;
      db.prepare(jobSql, (err, jobStatement) => {
        if (err) {
          db.disconnect();
          removeConn();
          return onDone(err);
        }

        const getMessages = (cb) => {
          const execCb = (err, rows) => {
            if (err) return cb(err);
            if (rows && rows.length > 0) {
              logfn(rows);
              this.last_row_id = rows[rows.length - 1].ROW_ID;
              Array.prototype.push.apply(this.all_messages, rows);
            }
            cb(null);
          };
          if (isHdb) {
            msgStatement.exec([requestId, this.last_row_id], {}, execCb);
          } else {
            msgStatement.exec([requestId, this.last_row_id], execCb);
          }
        };

        const checkJob = (cb) => {
          const execCb = (err, rows) => {
            if (err) return cb(err);
            if (!rows || rows.length === 0) return cb(null, 'RUNNING');
            const statuses = rows.map(r => (r.STATUS || '').toUpperCase());
            logger.trace(`Async make M_JOBS statuses for request ${requestId}: [${statuses.join(', ')}]`);
            /*
             * M_JOBS can have multiple rows for one REQUEST_ID (one per sub-job).
             * HANA uses many intermediate status values (e.g. 'COLLECTING MAKE FILES',
             * 'POST COMMIT STEPS', 'RUNNING') that are all in-progress.
             * Rather than maintaining an allow-list of in-progress statuses, we keep
             * polling until ALL rows are in a known terminal state.
             * 'ROLLED BACK' is ambiguous: it means either a genuine failure (changes reversed)
             * or a no-op (nothing to commit). The caller uses message SEVERITY to distinguish.
             */
            const allTerminal = statuses.every(
              s => TERMINAL_SUCCESS.has(s) || TERMINAL_ROLLBACK.has(s)
            );
            if (!allTerminal) {
              return cb(null, 'RUNNING');
            }
            const allSuccess = statuses.every(s => TERMINAL_SUCCESS.has(s));
            cb(null, allSuccess ? 'FINISHED' : 'ROLLED_BACK');
          };
          if (isHdb) {
            jobStatement.exec([requestId], {}, execCb);
          } else {
            jobStatement.exec([requestId], execCb);
          }
        };

        const poll = () => {
          getMessages((err) => {
            if (err) {
              db.disconnect();
              removeConn();
              return onDone(err);
            }
            checkJob((err, status) => {
              if (err) {
                db.disconnect();
                removeConn();
                return onDone(err);
              }
              if (status === 'RUNNING') {
                setTimeout(poll, 1000);
              } else {
                /*
                 * Final sweep: getMessages logs and appends remaining rows, then last_row_id is
                 * current — tasks.js filter(ROW_ID > last_row_id) returns empty (already logged).
                 * ROLLED_BACK is ambiguous: a genuine failure has SEVERITY='ERROR' messages;
                 * a no-op rollback ("No make necessary") has none.
                 */
                getMessages((err) => {
                  db.disconnect();
                  removeConn();
                  if (err) return onDone(err);
                  let rc;
                  // 'FINISHED' means all success terminals (FINISHED, COMMITTED, SIMULATED) — checkJob normalizes them
                  if (status === 'FINISHED') {
                    rc = 0;
                  } else {
                    rc = this.all_messages.some(m => m.SEVERITY === 'ERROR') ? -1 : 0;
                  }
                  onDone(null, {rc, messages: this.all_messages, last_row_id: this.last_row_id});
                });
              }
            });
          });
        };

        poll();
      });
    });
  }
}

/**
 * Class that wraps the make call to keep our state information etc "close"
 *
 * @class Make_Task
 */
class Make_Task {
  /**
   * Creates an instance of Make_Task.
   * @param {any} connection_id
   * @param {any} container
   * @param {any} hdiCreds
   * @param {any} schema
   * @param {any} options
   *
   * @memberOf Make_Task
   */
  constructor (connection_id, container, hdiCreds, schema, options = {}) {
    this.container = container;
    this.options = options;
    if (options.asyncMake) {
      this.asyncPoller = new AsyncPoller(schema, hdiCreds, options.asyncMakeTimeout);
    } else {
      this.poller = new Poller(connection_id, schema, hdiCreds);
    }
  }

  /**
   * Run the make, call the callback with errors, results and the last row id
   *
   * @param {any} callback
   *
   * @memberOf Make_Task
   */
  make (deployFiles, undeployFiles, pathParameters, parameters, callback) {
    if (this.options.asyncMake) {
      /*
       * simulate-make takes effect via the SIMULATE_MAKE=TRUE parameter in deployParameters,
       * which makeAsync receives and honors server-side — no client-side routing needed.
       */
      this.container.makeAsync(deployFiles, undeployFiles, pathParameters, parameters, (e, result) => {
        if (e) return callback(e);
        if (result.rc < 0) {
          return callback(null, {results: {rc: result.rc, messages: result.messages || []}, last_row_id: 0});
        }
        this.asyncPoller.poll(result.requestId, (err, pollResult) => {
          if (err) return callback(err);
          const {rc, messages, last_row_id} = pollResult;
          callback(null, {results: {rc, messages}, last_row_id});
        });
      });
    } else {
      this.container.make(deployFiles, undeployFiles, pathParameters, parameters, (e, results) => {
        const last_row_id = this.poller.stop_polling();
        callback(e, {results, last_row_id});
      });
      this.poller.start_polling();
    }
  }
}

module.exports = Make_Task;
