// @sap/cds-dk/lib/import/odata/annotationParser.js
// annotation parsing funtions common for v2parser.js and v4parser.js

"use strict";

const { warn } = require("../../util/term");

const known_vocabularies = {
	'Org.OData.Authorization.V1': 'Authorization',
	'Org.OData.Aggregation.V1': 'Aggregation',
	'Org.OData.Core.V1': 'Core',
	'Org.OData.Capabilities.V1': 'Capabilities',
	'Org.OData.Validation.V1': 'Validation',
	'Org.OData.Measures.V1': 'Measures',
	'Org.OData.JSON.V1': 'JSON',
	'Org.OData.Repeatability.V1': 'Repeatability',

	'com.sap.vocabularies.Analytics.v1': 'Analytics',
	'com.sap.vocabularies.CodeList.v1': 'CodeList',
	'com.sap.vocabularies.Common.v1': 'Common',
	'com.sap.vocabularies.Communication.v1': 'Communication',
	'com.sap.vocabularies.DataIntegration.v1': 'DataIntegration',
	'com.sap.vocabularies.Graph.v1': 'Graph',
	'com.sap.vocabularies.Hierarchy.v1': 'Hierarchy',
	'com.sap.vocabularies.HTML5.v1': 'HTML5',
	'com.sap.vocabularies.ODM.v1': 'ODM',
	'com.sap.vocabularies.Offline.v1': 'Offline',
	'com.sap.vocabularies.PDF.v1': 'PDF',
	'com.sap.vocabularies.PersonalData.v1': 'PersonalData',
	'com.sap.vocabularies.Session.v1': 'Session',
	'com.sap.vocabularies.UI.v1': 'UI'
};

function _replaceSpecialCharacters(text) {
	return text.replace(/(?:\\[rn]|[\r\n]+)+/gm, "\n").replace(/\s+/g, ' ').replace(/"/g, "&quot;").trim();
}

function captureAnnotationInCSN(targetString, annotationObject, csnDefsTargetObject) {
	/**
	 * For scenarios like:
	 * <Annotation Term="Core.Computed" />
	 * Here, 'annotationObject' would be null but in CSN we need to treat it as:
	 * @Core.Computed : true
	 */
	if (annotationObject === null) {
		annotationObject = true;
	}
	/**
	 * if annotationObject is empty list or empty object, we won't capture it.
	 * eg., <Annotation Term="Core.OptimisticConcurrency"> <Collection /> </Annotation>
	 * TODO: we might have to revisit it later for nested scenario.
	 */
	if ((annotationObject instanceof Array && annotationObject.length === 0) ||
		(annotationObject instanceof Object && Object.keys(annotationObject).length === 0)
	) {
		return;
	}
	/**
	 * <Annotation Term="Common.Text" Path="name"/> --> annotationTermObject = { "=": annotationTermObject.$Path };
	 */
	if (annotationObject.$Path) {
		annotationObject = { "=": annotationObject.$Path };
	}

	/**
	 * Annotations need to be flattened in CSN. E.g.
	 *
	 * @Core.ContentDisposition : {
	 *  Type    : 'inline',
	 *  Filename: typedImageFileName
	 * }
	 *
	 * is represented as:
	 *
	 * @Core.ContentDisposition.Type : 'inline',
	 * @Core.ContentDisposition.Filename : {'=': 'typedImageFileName'},
	 */
	if (annotationObject instanceof Object && !(annotationObject instanceof Array) && !(Object.keys(annotationObject).length === 1 && (Object.keys(annotationObject)[0] === '=' || Object.keys(annotationObject)[0] === 'xpr' || Object.keys(annotationObject)[0] === '#'))) {
		Object.keys(annotationObject).forEach(innerTargetString => {
			captureAnnotationInCSN(targetString + '.' + innerTargetString, annotationObject[innerTargetString], csnDefsTargetObject);
		});
	} else {
		csnDefsTargetObject[targetString] = annotationObject;
	}
}

function flattenAnnotationObject(annotationPrefix, annotationObject) {
	let flattenedResult = [];

	Object.entries(annotationObject).forEach(annotation => {
		let currentResult = [];
		/**
		 * if annotation looks like:
		 * { "#" : <value> } or { "=" : <value> }
		 * don't flatten because it is for the `EnumMember` & `Path` use cases respectively
		 */
		if (annotation[0] === '#' || annotation[0] === '=') {
			currentResult[0] = annotationPrefix;
			currentResult[1] = annotationObject;
		} else {
			currentResult[0] = annotationPrefix + '.' + annotation[0];
			currentResult[1] = annotation[1];
		}
		flattenedResult.push(currentResult);
	});

	return flattenedResult;
}

function removeTypeInAnnotation(annotationObject) {
	Object.keys(annotationObject).forEach(key => {
		if (key === "$Type") delete annotationObject[key];
		else if (annotationObject[key] instanceof Object) {
			removeTypeInAnnotation(annotationObject[key]);
		}
	});
}

function parseCoreTerm(annotationTerm, annotationTermObject, csnDefsTargetObject) {
	if (annotationTerm === "@Core.Description" || annotationTerm === "@Core.LongDescription") {
		// we don't capture `null` value for doc
		if (annotationTermObject === null) return;

		annotationTermObject = _replaceSpecialCharacters(annotationTermObject);

		if (csnDefsTargetObject.doc) {
			if (annotationTerm === "@Core.Description") {
				annotationTermObject = annotationTermObject + '\n\n' + csnDefsTargetObject.doc;
			} else {
				annotationTermObject = csnDefsTargetObject.doc + '\n\n' + annotationTermObject;
			}
		}
		return captureAnnotationInCSN('doc', annotationTermObject, csnDefsTargetObject);
	}

	/**
	 * <Annotation Term="Core.XYZ"/>
	 * For such scenario, mark the `annotationTermObject` as `true`
	 */
	if (annotationTermObject === null) annotationTermObject = true;

	if (annotationTerm === "@Core.OptionalParameter") {
		if (annotationTermObject && annotationTermObject.DefaultValue !== undefined) {
			csnDefsTargetObject.default = { val: annotationTermObject.DefaultValue };
		}
	} else {
		captureAnnotationInCSN(annotationTerm, annotationTermObject, csnDefsTargetObject);
	}
}

function parseCapabilitiesTerm(annotationTerm, annotationTermObject, csnDefsTargetObject) {
	let targetString;
	/**
	 * <Annotation Term="Capabilities.KeyAsSegmentSupported" />
	 * For such scenario, mark the `annotationTermObject` as `true`
	 */
	if (annotationTermObject === null) annotationTermObject = true;

	removeTypeInAnnotation(annotationTermObject);
	/**
	 * For entries like:
	 * <Annotation Term="Capabilities.FilterFunctions"> <Collection> <String>eq</String> <String>ne</String> </Collection> </Annotation>
	 * <Annotation Term="Capabilities.DeleteRestrictions.Deletable" Bool="true"/>
	 *
	 * simply preserve the value as it is
	 */
	if (annotationTermObject instanceof Array || !(annotationTermObject instanceof Object)) {
		captureAnnotationInCSN(annotationTerm, annotationTermObject, csnDefsTargetObject);
	}
	else {
		Object.keys(annotationTermObject).forEach(annoKey => {
			let currentAnnoValue = annotationTermObject[annoKey];
			// if the currentAnnoValue is an Object, we need to flatten it
			if (currentAnnoValue instanceof Object && !(currentAnnoValue instanceof Array)) {
				targetString = annotationTerm + '.' + annoKey;
				let flattenedResultList = flattenAnnotationObject(targetString, currentAnnoValue);

				for (let idx = 0; idx < flattenedResultList.length; idx++) {
					captureAnnotationInCSN(flattenedResultList[idx][0], flattenedResultList[idx][1], csnDefsTargetObject);
				}
			}
			// if the currentAnnoValue is a list or direct value, we preserve as it is
			else {
				targetString = annotationTerm + '.' + annoKey;
				captureAnnotationInCSN(targetString, currentAnnoValue, csnDefsTargetObject);
			}
		});
	}
}

function parseValidationTerm(annotationTerm, annotationTermObject, csnDefsTargetObject) {
	const resolveCastFor = ["Minimum", "Maximum", "MaxItems"];
	const term = annotationTerm.split('.')[1];

	if (resolveCastFor.includes(term) && annotationTermObject.$Cast) {
		annotationTermObject = Number(annotationTermObject.$Cast);
	}

	captureAnnotationInCSN(annotationTerm, annotationTermObject, csnDefsTargetObject);
}

/**
 * Capture all the annotation cases within an Element
 *
 * UnKnown term: <Annotation Term="Org.OData.ASDF.V1.qwe"/>
 * Known term: <Annotation Term="Org.OData.Measures.V1.Unit" String="Centimeters"/>
 *
 * if "--include-namespaces *" then we also capture for ABS and QWE
 * <TypeDefinition Name="TD_2" UnderlyingType="Edm.Int32" ABS:asdf="testing" QWE:qwe="testing"/>
 */
function parseAnnotationTerms(annotationTerm, annotationTermObject, csnDefsTargetObject) {
	const term = annotationTerm.split('.');
	const vocabAlias = term[0];

	switch (vocabAlias) {
		case "@Core":
			parseCoreTerm(annotationTerm, annotationTermObject, csnDefsTargetObject);
			break;
		case "@Capabilities":
			parseCapabilitiesTerm(annotationTerm, annotationTermObject, csnDefsTargetObject);
			break;
		case "@Validation":
			parseValidationTerm(annotationTerm, annotationTermObject, csnDefsTargetObject);
			break;
		default:
			captureAnnotationInCSN(annotationTerm, annotationTermObject, csnDefsTargetObject);
			break;
	}
}

function validateAnnotationNamespace(annotationTerm, vocabNamespaceToAlias) {
	let aliasOrNamespace = annotationTerm.substring(1, annotationTerm.lastIndexOf('.'));

	// This check is for ignoring the @cds annotations given for special cases, e.g. @cds.validate = false for collided action/function
	if (aliasOrNamespace === "cds") {
		return;
	}

	for (let alias of Object.values(known_vocabularies)) {
		if (alias === aliasOrNamespace) {
			return;
		}
	}

	for (let [namespace, alias] of Object.entries(vocabNamespaceToAlias || {})) {
		if (alias === aliasOrNamespace || namespace === aliasOrNamespace) {
			console.log(warn(annotationTerm + " AnnotationTerm belongs to an unknown namespace: " + namespace));
			return;
		}
	}

	console.log(warn(annotationTerm + ": Reference of this AnnotationTerm is not present in the edmx file."));
}

function resolveAnnotationTerm(schemaDataAnnotation, parserContext) {
	for (const [namespace, alias] of Object.entries(parserContext.vocabNamespaceToAlias || {})) {
		if (known_vocabularies[namespace]) {
			// Case 1: Namespace belongs to known vocabularies. Replacing the namespace/alias of the doc with standard alias
			if (schemaDataAnnotation.includes('@' + alias + '.')) {
				schemaDataAnnotation = schemaDataAnnotation.split('@' + alias + '.').join('@' + known_vocabularies[namespace] + '.');
			}
			if (schemaDataAnnotation.includes('@' + namespace + '.')) {
				schemaDataAnnotation = schemaDataAnnotation.split('@' + namespace + '.').join('@' + known_vocabularies[namespace] + '.');
			}
		} else {
			// Case 2: Namespace doesn't belong to known vocabularies. Replacing the namespace with alias(if provided) in the doc.
			if (schemaDataAnnotation.includes('@' + namespace + '.') && alias) {
				schemaDataAnnotation = schemaDataAnnotation.split('@' + namespace + '.').join('@' + alias + '.');
			}
		}
	}
	/**
	 * Case 3: Namespace doesn't have edmx:Reference. But it is present in known_vocabularies.
	 *         Replacing it with standard alias.
	 */
	for (const [namespace, alias] of Object.entries(known_vocabularies)) {
		if (schemaDataAnnotation.includes(namespace)) {
			schemaDataAnnotation = schemaDataAnnotation.split(namespace).join(alias);
		}
	}

	let index = schemaDataAnnotation.indexOf('"@');
	while (index != -1) {
		let i = schemaDataAnnotation.indexOf('"', index + 1);
		let annotationTerm = schemaDataAnnotation.substring(index + 1, i);
		validateAnnotationNamespace(annotationTerm, parserContext.vocabNamespaceToAlias);
		index = schemaDataAnnotation.indexOf('"@', i + 1);
	}

	return JSON.parse(schemaDataAnnotation);
}

function parseAnnotations(schemaAnnotations, csnDefs, parserContext, resolveTarget) {
	Object.keys(schemaAnnotations || {}).forEach(annotationTargetPath => {
		let csnDefsTargetObject = resolveTarget(annotationTargetPath, csnDefs, parserContext);
		if (csnDefsTargetObject) {
			let annotationTerms = schemaAnnotations[annotationTargetPath];
			Object.keys(annotationTerms).forEach(annotationTerm => {
				let annotationTermObject = schemaAnnotations[annotationTargetPath][annotationTerm];
				parseAnnotationTerms(annotationTerm, annotationTermObject, csnDefsTargetObject);
			});
		}
	});
}

module.exports = {
	captureAnnotationInCSN,
	flattenAnnotationObject,
	removeTypeInAnnotation,
	parseCoreTerm,
	parseCapabilitiesTerm,
	parseValidationTerm,
	parseAnnotationTerms,
	validateAnnotationNamespace,
	resolveAnnotationTerm,
	parseAnnotations
};
